### PR TITLE
RFD 16: Dynamic Configuration

### PR DESCRIPTION
**Pull request by andrejtokarcik**
The PR #5098 revealed several issues in connection with the goal of providing support for user-managed resource-based dynamic configuration. This RFD presents the open questions in the form of practical branching scenarios. For now it should aid in the decision making process, eventually it would serve to clarify the agreed-upon design.

Also related: #4867, #4667, #4774.

### CODE REVIEW

**Comment by russjones** [comment]
@andrejtokarcik It's a little hard to give line-by-line feedback here, because so many scenarios exists in the RFD. Below is the behavior that I think should exist. @klizhentas @alex-kovoy @benarent what do you two think about the following behavior?

### Description

If an `authentication` block in specified in `teleport.yaml`, upon startup Teleport should create and apply the resource to the backend.

If an `authentication` block is not specified in `teleport.yaml`, upon startup, Teleport should do nothing.

When getting the resource, if the backend returns not found, a sane default resource should be returned similar to [DefaultClusterConfig](https://github.com/gravitational/teleport/blob/master/lib/services/clusterconfig.go#L126-L144).

When setting the resource the function should be updated to take a force flag.

1. Without the force flag, the function should check if the resource is defined before setting it. If the resource is defined, an error should be returned to the user informing them that the resource is already configured either in file configuration or though a resource.
2. With the force flag, override and apply whatever the user is requesting.

### Examples

1. For current users that define `authentication`, if they use `tctl create` they will get an error message saying that this resource is already defined. If they use `tctl create -f` they can override the resource until the next restart.

2. If `authentication` is not defined and `tctl create` is never used, Teleport will always return a `services.DefaultAuthPreference`.

3. If `authentication` is not defined and `tctl create` is used, it will apply the first time, after an error message saying that this resource is already defined will be returned. If they use `tctl create -f` they can always set the resource.

**Comment by fspmarshall** [comment]
> If an `authentication` block in specified in `teleport.yaml`, upon startup Teleport should create and apply the resource to the backend.
> 
> If an `authentication` block is not specified in `teleport.yaml`, upon startup, Teleport should do nothing.

Not updating the backend if nothing is specified is problematic since a user who did not dynamically apply a configuration value would expect that removing that value from their configuration file would revert to the default behavior.  It might be simpler to track both the static and dynamic configuration separately, and then merge them as needed.  That way we don't have to worry about weird ordering effects.

**Comment by russjones** [comment]
@fspmarshall You're proposing something like the following?

```
if dynamic && static:
   return static
elif !dynamic && static:
   return static
elif dynamic && !static:
  return dynamic
elif !dynamic && !static:
   return default
```

**Comment by fspmarshall** [comment]
@russjones Yes, thats the state-machine.  More simply (but equivalent):

```
if static:
    return static
elif dynamic:
    return dynamic
else:
    return default
```

File: rfd/0016-dynamic-configuration.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+2. The auth server initialization procedure MUST commit data derived from the
+   static configuration to the backend, overwriting any potential resource data
+   already stored there.
+
+3. `tctl get cap` returns the resource data in line with `teleport.yaml`.
+
+Therefore, static configuration is always given precedence.  This should also
+ensure backward compatibility with the already established workflows.
+
+### Scenario 2: static never specified & dynamic not updated by user
```
Comment (klizhentas) [inline rfd/0016-dynamic-configuration.md:56]: I really like the decision tree layout in this RFD @andrejtokarcik
Really helps to  describe all possible permutations and edge cases

**Comment by andrejtokarcik** [comment]
@russjones You seem to be simply proposing the following combination of the RFD Choices: 2.B, 3.A, 5.A, 6.A.

@fspmarshall's remark appears to amount to Choice 3.B of Scenario 3. This would also bear upon the last table in the RFD, making it equivalent to the mentioned state machines. However, this still seems to be the most troublesome/ambiguous case especially in connection with 6.A: if a `teleport.yaml` section is present, then its resource updated with `tctl create -f`, then the section is removed, what configuration should be restored upon restart: default or the updated resource?

At this point I believe @fspmarshall would object to Choice 6.A altogether as he's already [expressed such concerns](https://github.com/gravitational/teleport/pull/5098#pullrequestreview-550527543).

**Comment by klizhentas** [comment]
I would like to avoid having two different resources and storages for static and dynamic configuration. This will make it really hard to use the API.

@andrejtokarcik thank you for describing all the edge-cases in such detail.

Soon, we are going to add [includes](https://github.com/gravitational/teleport/issues/4997). That is why I would like to consider our static configuration as another user of our resources.

It means that there is no difference between static or dynamic configuration resource, there is only one resource. Just as with other clients, the last write wins.

In practice, it means that:

**Scenario 1 (Static configuration specified).**
Static configuration will overwrite resources on restarts.

This is our first major use case - configuration is managed by users via static configuration only.

**Scenario 2: static never specified & dynamic not updated by user**

If the server starts and finds no resource, it will create a default one. (Choice 2B.). 
When creating such a resource, the server will add a label in the resource:

```
metadata:
   label: 
      source: config-file
      server-id: id of the server that created the value
```

It will use this label to resolve edge cases below. It will also make the system more stable, because it will enforce
the invariant - there is going to be some resource present after first server start. It is important to make API usage convenient - there is not going to be a situation when there is no resource for API users to consume.

** Scenario 3: static reverted to unspecified & dynamic not updated by user**

This is an edge case when users switch to static to dynamic configuration and vice versa.

The server (re)starts and finds an empty config with no parameters in it's static section. It sees the resource with `source: config-file`.

* It sees it's own server ID and `source: config-file` in the labels. it overwrites the values to default values.
* It sees other server ID and `source: config-file` in the labels. it overwrites the values to default values and logs a warning saying that there was another server with different config.
* It sees that there is no `source: config-file` in the labels, it does nothing

So it's a combination of 3.A if the existing value is managed by config files and 3.B will get overwritten if the source of the resource is config files.

**Scenario 4: static never specified & first `tctl create`**
** Scenario 5: static never specified & repeated `tctl create

These scenarios describe the second most popular use-case - resources are managed completely dynamically.
We should follow `Choice 5.A` in case of `tctl create` as it's consistent with the way the resources are specified `create` will return already exists error and `create -f` will overwrite the existing resource if it exists.

### Scenario 6: static specified & `tctl create` attempt
It's a great catch @andrejtokarcik.

At first, I'd follow `Choice 6.B` and issue the following message:

```
This resource was managed by static configuration. If there are servers with static configuration, this could be overwritten by configs. We recommend removing configuration from teleport.yaml, restarting the servers and trying this command again.

If you would still like to proceed, re-run the comand with confirm flag:

$ tctl create -f authpref.yaml --confirm
```

So users will proceed to choice `6.A` in case if they would like to test something.

**Comment by andrejtokarcik** [comment]
@klizhentas Thanks for your thorough comment.

Is my understanding correct that under Scenario 1, the `source: config-file` label should be set as well?

The idea of using labels in this way is intriguing but I'm worried about the possibility of these labels being set accidentally by a well-intentioned user, since `tctl create -f` overwrites not only the `spec` part but also the `metadata` part of a resource. In particular, let's say one performs the usual steps of `tctl get`, edit some `spec` fields in the output YAML, `tctl create -f` with the edited YAML, then they'll end up with a config that's still labelled as `source: config-file`, thereby breaking our logic based on distinguishing the configuration source by the label.

In Scenario 3, since the dynamic configuration is assumed to have not been updated by the user, `source: config-file` is necessarily present and your third bullet point never comes into play? Also, (given at least two servers) isn't the check for `server-id` bound to emit the warning *ad infinitum* as every server getting online finds `server-id` of another server and immediately overwrites it with its own `server-id`?

There is currently no `--confirm` flag so the idea is to introduce it just because of Scenario 6? How should it behave with other resources? Wouldn't it be simpler to just return the first line of the proposed message seeing that this is probably just a minority use case of volatile/temporary experimenting with the config settings?

**Comment by klizhentas** [comment]
> @klizhentas Thanks for your thorough comment.
> 
> Is my understanding correct that under Scenario 1, the `source: config-file` label should be set as well?
> 

Seems correct - every time the system sets the value it will mark the system.

> The idea of using labels in this way is intriguing but I'm worried about the possibility of these labels being set accidentally by a well-intentioned user, since `tctl create -f` overwrites not only the `spec` part but also the `metadata` part of a resource. In particular, let's say one performs the usual steps of `tctl get`, edit some `spec` fields in the output YAML, `tctl create -f` with the edited YAML, then they'll end up with a config that's still labelled as `source: config-file`, thereby breaking our logic based on distinguishing the configuration source by the label.
> 

That's the trade-off I am OK to make. If user overwrites the labels set by the system, they may break the system. Kubernetes follows the similar approach by marking owner reference in the object controller creates. They set system labels and document their usage. If you update the system labels, the system may misbehave. 

> In Scenario 3, the dynamic configuration is assumed to have not been updated by the user so `source: config-file` is necessarily present and your third bullet point would actually never come into play, or would it? 
> 

Scenario 3 that you described says: 

> The `authentication` section used to be specified in `teleport.yaml` but
>   it has been removed and left unspecified since the last auth server init.
>   No user changes have been made to the dynamic configuration.

So I was assuming that there is existing resource then with `source: config-file` because it was used to be specified. Once it's removed one of the auth server restarts and observes the resource then. 

> Also, (given at least two servers) isn't the check for `server-id` bound to emit the warning _ad infinitum_ as every server getting online finds `server-id` of another server and immediately overwrites it with its own `server-id`?

I was thinking to emit the warning with `server-id` only if configuration present in the resource is different from what this server is going to write. But it could be a source of confusion, yes, so let's skip it for now.

> There is currently no `--confirm` flag so the idea is to introduce it just because of Scenario 6? How should it behave with other resources? Wouldn't it be simpler to just return the first line of the proposed message seeing that this is probably just a minority use case of volatile/temporary experimenting with the config values?

I've seen `--confirm` flag used in other projects, like terraform used for `terraform destroy` and other potentially dangerous operations, so I have plans for it outside of the scope of this issue :) We can introduce it now anduse it more in the future.

**Comment by andrejtokarcik** [comment]
> If user overwrites the labels set by the system, they may break the system.

There is no intentional overwriting or updating of the labels in my "well-intentioned user" example though. To be more precise:
1. A user intends to switch over to dynamic configuration. The relevant config section is already removed from `teleport.yaml`.
2. He runs `tctl get` to fetch the default configuration with the idea of using the resulting YAML as a stepping stone because (as he correctly assumes) to write a resource YAML from scratch just isn't right.
3. He updates some `spec` values in the YAML, not noticing there is a `metadata.source` field. And even if he did, `source: config-file` may sound good/relevant enough so as not to urge him to get rid of it.
4. He runs `tctl create` expecting his changes to survive a restart which is however not the case because this would be evaluated as Scenario 3 (i.e., a `source: config-file` configuration to be reverted back to default).

In my view, it'd be more feasible to support this straightforward way of switching to dynamic configuration directly (because using `tctl create` should be a sufficient indication of this intention) instead of having to put warnings everywhere not to forget to remove the `metadata.source` field.

> So I was assuming that there is existing resource then with `source: config-file`

Exactly - so the condition of your third bullet point ("there is no `source: config-file` in the labels") would never apply, right?

**Comment by klizhentas** [comment]
> > If user overwrites the labels set by the system, they may break the system.
> 
> There is no intentional overwriting or updating of the labels in my "well-intentioned user" example though. To be more precise:
> 
> 1. A user intends to switch over to dynamic configuration. The relevant config section is already removed from `teleport.yaml`.
> 2. He runs `tctl get` to fetch the default configuration with the idea of using the resulting YAML as a stepping stone because (as he correctly assumes) to write a resource YAML from scratch just isn't right.
> 3. He updates some `spec` values in the YAML, not noticing there is a `metadata.source` field. And even if he did, `source: config-file` may sound good/relevant enough so as not to urge him to get rid of it.
> 4. He runs `tctl create` expecting his changes to survive a restart which is however not the case because this would be evaluated as Scenario 3 (i.e., a `source: config-file` configuration to be reverted back to default).
> 
> In my view, it'd be more feasible to support this straightforward way of switching to dynamic configuration directly (because using `tctl create` should be a sufficient indication of this intention) instead of having to put warnings everywhere not to forget to remove the `metadata.source` field.
> 

This is one of those cases where it's impossible to have all edge-cases work perfectly well without sacrificing U.X. for other use cases. Because this is not a priority use case and is an edge case "User is switching from static to dynamic". I'm OK making this design tradeoff. I'd just issue a warning when auth server overwrites the config set by user (resource has no `source: config-file`). We should prioritize use cases when users are either using static config files only "hobbyist, small scale deployment" or using completely dynamic configuration with minimal static config file "automate everything trhough reources/terraform" one. I will write a guide on migration from one users to another.

> > So I was assuming that there is existing resource then with `source: config-file`
> 
> Exactly - so the condition of your third bullet point ("there is no `source: config-file` in the labels") would never apply, right? 

Ah, regarding the third bullet point, sure it won't apply in this case.

**Comment by andrejtokarcik** [comment]
The following question occurred to me while working on a basic implementation:

Does the issue of data structure updates come into play? For instance, aren't the structures stored in the backend currently simply assumed to get replaced with their newer/extended counterparts when an auth server starts up after a Teleport version upgrade? Is it safe to introduce the dynamic logic with the effect that the auth server is not guaranteed to overwrite the config resources stored in the backend? Isn't there a risk of indefinitely retaining stale/incompatible data structures in the backend (those committed dynamically using `tctl create`)?

**Comment by klizhentas** [comment]
@andrejtokarcik it is safe to use the data structures. They are versioned and when we create new versions we do migrations and compatibility on the fly. For example we will support v2 and v3 versions of resources for a while and then switch to just v3 after a couple of versions.

**Comment by klizhentas** [comment]
see for example users: 

https://github.com/gravitational/teleport/blob/master/lib/services/user.go#L582

**Comment by andrejtokarcik** [comment]
Updated the RFD with a new section on implementation. It basically captures the conclusions of our discussion above with the following differences:

1. Used `origin` as the label name instead of `source` mainly because saying "resource source" sounds kind of awkward.
2. The proposed `server-id` label and the related logic is not discussed at all.
2. Indicated the possible usefulness of the `tctl rm` operation.

@klizhentas If you could have a look, I'd be very grateful to know if I've understood your ideas correctly. If there are no major issues I'll open a PR with the actual code tomorrow.

**Comment by andrejtokarcik** [comment]
I forgot to mention that one more change I had to made was to use `origin: defaults` instead of `origin: source-file` for Scenario 2. This is needed so that `tctl create` can determine when it's allowed to overwrite the resource stored in the backend.

**Review by klizhentas — APPROVED** [review]

**Comment by klizhentas** [comment]
@andrejtokarcik @russjones is this getting in 6.0 or 6.1?

**Pull request merged**